broker/src/main/java/io/zeebe/broker/clustering/base/topology/TopologyManagerImpl.java

```diff
@@ -125,45 +112,42 @@ public void event(ClusterMembershipEvent clusterMembershipEvent) {
 
   // Remove a member from the topology
   private void onMemberRemoved(BrokerInfo brokerInfo) {
-    final NodeInfo nodeInfo = topology.getMember(brokerInfo.getNodeId());
-    if (nodeInfo != null) {
-      topology.removeMember(nodeInfo);
-      notifyMemberRemoved(nodeInfo);
-    }
+    brokerInfo.consumePartitions(
+        partition -> removeIfLeader(brokerInfo, partition),
+        (leaderPartitionId, term) -> {},
+        followerPartitionId -> {});
   }
 
-  // Add a new member to the topology, including its interface's addresses
-  private void onMemberAdded(BrokerInfo brokerInfo) {
-    final String commandApiAddress = brokerInfo.getCommandApiAddress();
-    if (commandApiAddress != null) {
-      final NodeInfo nodeInfo =
-          new NodeInfo(brokerInfo.getNodeId(), SocketAddress.from(commandApiAddress));
-
-      if (topology.addMember(nodeInfo)) {
-        notifyMemberAdded(nodeInfo);
-      }
-    } else {
-      LOG.warn(
-          "Ignoring broker info from node id {} as no command API address is present",
-          brokerInfo.getNodeId());
+  private void removeIfLeader(BrokerInfo brokerInfo, Integer partition) {
+    final BrokerInfo currentLeader = partitionLeaders.get(partition);
+    if (currentLeader != null && currentLeader.getNodeId() == brokerInfo.getNodeId()) {
+      partitionLeaders.remove(partition);
     }
   }
 
   // Update local knowledge about the partitions of remote node
   private void onMetadataChanged(BrokerInfo brokerInfo) {
-    final NodeInfo nodeInfo = topology.getMember(brokerInfo.getNodeId());
-
     brokerInfo.consumePartitions(
-        leaderPartitionId -> {
-          topology.updatePartition(leaderPartitionId, nodeInfo, RaftState.LEADER);
-          notifyPartitionUpdated(leaderPartitionId, nodeInfo);
+        (leaderPartitionId, term) -> {
+          if (updatePartitionLeader(brokerInfo, leaderPartitionId, term)) {
+            notifyPartitionLeaderUpdated(leaderPartitionId, brokerInfo);
+          }
         },
         followerPartitionId -> {
-          topology.updatePartition(followerPartitionId, nodeInfo, RaftState.FOLLOWER);
-          notifyPartitionUpdated(followerPartitionId, nodeInfo);
+          removeIfLeader(brokerInfo, followerPartitionId);
         });
   }
 
+  private boolean updatePartitionLeader(BrokerInfo brokerInfo, int leaderPartitionId, long term) {
+    final BrokerInfo currentLeader = partitionLeaders.get(leaderPartitionId);
+    if (currentLeader != null
+        && currentLeader.getPartitionLeaderTerms().get(leaderPartitionId) >= term) {
```

Comment:
I think we should use getOrDefault here right?

Reply:
If there is a leader, the term will be there right?

Reply:
If there is no bug I hope :sweat_smile:

Reply:
Isn't it then better to crash with a null pointer exception, than hiding it with a default value?

Reply:
Agree but NPE is normally not really meaningful and context free. So it is hard to determine what exactly is null in that case

Reply:
@Zelldon Would it be better if it is handled like this? https://github.com/zeebe-io/zeebe/pull/3375/commits/8467507998043517896e65c8b2c6c1f900d9365b